plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/LocalAppEnginePublishOperation.java

```diff
@@ -0,0 +1,117 @@
+import org.eclipse.core.runtime.IAdaptable;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.MultiStatus;
+import org.eclipse.wst.server.core.IModule;
+import org.eclipse.wst.server.core.IServer;
+import org.eclipse.wst.server.core.model.IModuleResource;
+import org.eclipse.wst.server.core.model.IModuleResourceDelta;
+import org.eclipse.wst.server.core.model.PublishOperation;
+import org.eclipse.wst.server.core.model.ServerBehaviourDelegate;
+import org.eclipse.wst.server.core.util.PublishHelper;
+
+import com.google.common.collect.Lists;
+
+/**
+ * Handles the publishing operations for the App Engine development server.
+ */
+public class LocalAppEnginePublishOperation extends PublishOperation {
+  private static final String PLUGIN_ID = LocalAppEnginePublishOperation.class.getName();
+
+  /**
+   * @throws {@link CoreException} if status list is not empty
+   */
+  private static void checkStatuses(List<IStatus> statusList) throws CoreException {
+    if (statusList == null || statusList.isEmpty()) {
+      return;
+    }
+    if (statusList.size() == 1) {
+      throw new CoreException(statusList.get(0));
+    }
+    IStatus[] children = statusList.toArray(new IStatus[statusList.size()]);
+    throw new CoreException(new MultiStatus(PLUGIN_ID, 0, children, "Error during publish operation", null));
+  }
+
+  private LocalAppEngineServerBehaviour server;
+  private IModule[] modules;
+  private int kind;
+  private int deltaKind;
+  private PublishHelper helper;
+
+  public int getKind() {
+    return REQUIRED;
+  }
+
+  @Override
+  public int getOrder() {
+    return 0;
+  }
+
+  /**
+   * Construct the operation object to publish the specified modules(s) to the
+   * specified server.
+   */
+  public LocalAppEnginePublishOperation(LocalAppEngineServerBehaviour server, int kind, IModule[] modules,
+      int deltaKind) {
+    super("Publish to server", "Publish modules to App Engine Development Server");
+    this.server = server;
+    this.kind = kind;
+    this.deltaKind = deltaKind;
+    IPath base = server.getRuntimeBaseDirectory();
+    helper = new PublishHelper(base.toFile());
+
+    if (modules != null) {
+      this.modules = Arrays.copyOf(modules, modules.length);
+    } else {
+      this.modules = new IModule[0];
+    }
+  }
+
+  @Override
+  public void execute(IProgressMonitor monitor, IAdaptable info) throws CoreException {
+    List<IStatus> statusList = Lists.newArrayList();
+    IPath deployPath = server.getModuleDeployDirectory(modules[0]);
+    publishDirectory(deployPath, statusList, monitor);
+    checkStatuses(statusList);
+    server.setModulePublishState2(modules, IServer.PUBLISH_STATE_NONE);
+  }
+
+  /**
+   * Publish modules as directory.
+   */
+  private void publishDirectory(IPath path, List<IStatus> statusList, IProgressMonitor monitor) throws CoreException {
```

Comment:
OK. Agreed we shouldn't make that change here then. Would it make sense to add a TODO/issue to move this to a MultisStatus instead of a List in another PR?